### DISCUSSION TITLE
Prediction --> Editable Instance is taking too long

### DISCUSSION THREAD

**Discussion by lisadiez** [talmolab/sleap#1123] [Help!]
Hi!

We have a project with 23,011 predicted frames and are starting to correct the predictions. We noticed the conversion from predicted instance to editable instance is taking a long time, as well as ID changes.

https://user-images.githubusercontent.com/116014547/213214448-b96436da-a447-451c-9314-64d6b27db3ed.mov

I'm not sure you can appreciate it in the screen recording, but it's taking around 3-5s after I double-click.

Is this because we have too many predicted frames?

**Comment by talmo** [comment]
Some preliminary analysis @roomrys and I did:

For a project with 1 video, 8 `Track`s, no `Suggestions`, ~270k `LabeledFrame`s, all with <=8 `PredictedInstance`s in each, we see a similar latency spike when creating a user `Instance` from a `PredictedInstance`.

On my PC, scrolling through the frames calls [`on_data_update`](https://github.com/talmolab/sleap/blob/31cb5a8c80b6c3fa6bb38f5e89322fd3e33c0bfb/sleap/gui/app.py#L1301) which runs in about ~1 ms. *Note:* Scrubbing by dragging on the seekbar does NOT set the update topics to trigger [`self.plotFrame()`](https://github.com/talmolab/sleap/blob/31cb5a8c80b6c3fa6bb38f5e89322fd3e33c0bfb/sleap/gui/app.py#L1318) -- this is handled via the `frame_idx` context signal rather than in [`on_data_update`](https://github.com/talmolab/sleap/blob/31cb5a8c80b6c3fa6bb38f5e89322fd3e33c0bfb/sleap/gui/app.py#L1301).

Ok now, creating an `Instance` from a `PredictedInstance` is quite slow. This calls the [`AddInstance`](https://github.com/talmolab/sleap/blob/31cb5a8c80b6c3fa6bb38f5e89322fd3e33c0bfb/sleap/gui/commands.py#L2620) command, which sets the `UpdateTopic.frame, UpdateTopic.project_instances, UpdateTopic.suggestions` topics for updating.

Doing some profiling, reveals that this part of [`on_data_update`](https://github.com/talmolab/sleap/blob/31cb5a8c80b6c3fa6bb38f5e89322fd3e33c0bfb/sleap/gui/app.py#L1301) takes about **89 ms**:
https://github.com/talmolab/sleap/blob/31cb5a8c80b6c3fa6bb38f5e89322fd3e33c0bfb/sleap/gui/app.py#L1310-L1318
I'm guessing this is because we have to draw the yellow "reminder" boxes around the other instances? We can test this by profiling the lower level [`QtVideoPlayer.plot()`](https://github.com/talmolab/sleap/blob/31cb5a8c80b6c3fa6bb38f5e89322fd3e33c0bfb/sleap/gui/widgets/video.py#L470-L490) method to see if just rendering frames with user instances in these conditions is slow in general.

Meanwhile, here's the big culprit -- this part of [`on_data_update`](https://github.com/talmolab/sleap/blob/31cb5a8c80b6c3fa6bb38f5e89322fd3e33c0bfb/sleap/gui/app.py#L1301) takes **264 ms**:
https://github.com/talmolab/sleap/blob/31cb5a8c80b6c3fa6bb38f5e89322fd3e33c0bfb/sleap/gui/app.py#L1320-L1328

The rest takes <1ms and the total  [`on_data_update`](https://github.com/talmolab/sleap/blob/31cb5a8c80b6c3fa6bb38f5e89322fd3e33c0bfb/sleap/gui/app.py#L1301) takes **356 ms**.

Summarizing:

Total: 356 ms
`plotFrame`: 89 ms (25%)
`_update_seekbar_marks`: 264 ms (74%)

So these are the places we need to optimize. We'll probably get a big enough boost by just fixing how we render the seekbar.

We'll work on this and update this thread when we have a potential solution.

**Comment by catubc** [comment]
Thank you both, this is very helpful.  It wasn't completely clear why this scales with # of video frames!?

In any case, we will decimate the dataset to every 4th or 5th frame for now while we wait for a speedup.
